### ISSUE TITLE
More efficient encoding / decoding

### ISSUE THREAD

**Issue by dmjio** [valderman/selda#18]
We could probably use [postgresql-binary](https://hackage.haskell.org/package/postgresql-binary-0.12.1/docs/PostgreSQL-Binary-Decoding.html) as a drop-in replacement for what we're currently doing here: 
```haskell
 | t == boolType    = SqlBool $ readBool val
 | t == intType     = SqlInt $ read (BS.unpack val)
 | t == doubleType  = SqlFloat $ read (BS.unpack val)
 | t `elem` textish = SqlString (decodeUtf8 val)
```
This would also allow us to be more specific about `time` and other `textish` data.
```haskell
valueParser :: Value a -> ByteString -> Either Text a
```

Same story for `encoding` too:
```haskell
fromSqlValue :: Lit a -> Maybe (Oid, BS.ByteString, Format)
fromSqlValue (LitB b)    = Just (boolType, if b then "true" else "false", Text)
fromSqlValue (LitI n)    = Just (intType, BS.pack $ show n, Text)
fromSqlValue (LitD f)    = Just (doubleType, BS.pack $ show f, Text)
fromSqlValue (LitS s)    = Just (textType, encodeUtf8 s, Text)
fromSqlValue (LitTS s)   = Just (timestampType, encodeUtf8 s, Text)
fromSqlValue (LitTime s) = Just (timeType, encodeUtf8 s, Text)
fromSqlValue (LitDate s) = Just (dateType, encodeUtf8 s, Text)
fromSqlValue (LitNull)   = Nothing
fromSqlValue (LitJust x) = fromSqlValue x
```
w/ [`postgresql-binary`](https://hackage.haskell.org/package/postgresql-binary-0.12.1/docs/PostgreSQL-Binary-Encoding.html)
```haskell
PostgreSQL.Binary.Encoding.encodingBytes :: Encoding -> ByteString
```

Should probably get away from `[[SqlValue]]` as well, since that's probably not going to be the most efficient.

**Comment by valderman** [comment]
Yep, the performance of encoding/decoding is currently likely to be abysmal. `pack . show` is about as quick and dirty as it gets. To get away from the nested lists of `SqlValue`s, one option would be to simply have each backend marshal directly to the correct type. It'd make writing a new backend a bit more cumbersome, but it's not like that's a common task anyway. Any other ideas that don't incur this maintainability penalty while being reasonably performant?

**Comment by dmjio** [comment]
> one option would be to simply have each backend marshal directly to the correct type

I assume the binary formats could change across backends. Albeit cumbersome, this might be necessary for performance. We might need some kind of typeclass abstraction, yet I'm unsure how to introduce that without causing orphan instances. The class for decoding / encoding could live in `selda`, but the instances would live in the `sqlite/postgresql` backends.

**Comment by winterland1989** [comment]
I need proper support for prepared statements built in to selda `SeldaBackend`. MySQL only use binary format in prepared statements.

**Comment by valderman** [comment]
@winterland1989, this issue shouldn't affect MySQL support, since backends have full control over how they talk to the database engine. If would be helpful if you could create a separate issue about MySQL support, so we can collect the changes that need to be made (for instance, which parts of queries need to be parameterized) in one place.

**Comment by winterland1989** [comment]
@valderman AFAICT. I'm not being able to write a mysql backend because how query pretty-printter currently built [here](https://github.com/valderman/selda/blob/6a6650ac3604702649dc64252af65408e01eba24/selda/src/Database/Selda/SQL/Print.hs#L92):

```
-- | Pretty-print a literal as a named parameter and save the
--   name-value binding in the environment.
ppLit :: Lit a -> PP Text
ppLit LNull     = pure "NULL"
ppLit (LJust l) = ppLit l
ppLit l         = do
  PPState ps ts ns qns tr <- get
  put $ PPState (Param l : ps) ts (succ ns) qns tr
  return $ Text.pack ('$':show ns)
```

MySQL can't recognize this kind of variable placeholder, so i need a way to let backend custom how placeholder are pretty-printed.

As for prepared statement support, i mean current API of  `SeldaBackend` force i to use prepared statement in one go, but that defeated the whole purpose of prepared statement.

**Comment by valderman** [comment]
@winterland1989 I think there are two separate issues here, one for the requirements of MySQL support (#22), and one for allowing prepared statements to be used for performance, and not only for input hygiene (#23). There are probably more differences between MySQL and the current supported backends than just the different syntax of the parameter placeholder though. Please add any additional changes needed to #22.

**Comment by winterland1989** [comment]
Exactly!  I'll try to give more detail while hacking.

**Comment by valderman** [comment]
Fixed as of ecf57aea1ac90a14446596de14acd4e063a0c1a2.
